### PR TITLE
Experimental

### PR DESCRIPTION
**Pull request by leonardopetrini**
I'm still working on the signal_1d dataset

### CODE REVIEW

File: dynamics.py [hunk 1]
```diff
@@ -213,7 +213,7 @@ def train_kernel(ktrtr, ytr, tau, alpha, loss_prim, max_dgrad=math.inf, max_dout
     margin = 0
 
     lprim = loss_prim(otr, ytr)
-    grad = ktrtr @ lprim / len(ytr)
+    grad = ktrtr @ lprim    / len(ytr)
```
Comment (mariogeiger) [inline dynamics.py]: pourquoi ? ^^

File: dataset/__init__.py [hunk 2]
```diff
@@ -9,15 +9,27 @@
...
 import scipy.special
 import torch
+import torch.nn.functional as F
+
+def inverf2(x):
+    """ Inverse error function in 2d."""
+    return (-2 * (1 - x).log()).sqrt()
+
+
+def stretch(x, dims, factor):
```
Comment (mariogeiger) [inline dataset/__init__.py]: ne définit pas une fonction aussi simple pour ne l'utiliser qu'une seule fois

**Review by mariogeiger — COMMENTED** [review]
Dis moi quand tu veux que je merge

File: dynamics.py (hunk 1 above)
Comment (leonardopetrini) [reply dynamics.py]: typo en lisant le code !

File: dataset/__init__.py (hunk 2 above)
Comment (leonardopetrini) [reply dataset/__init__.py]: ook !

File: dataset/__init__.py [hunk 3]
```diff
@@ -168,11 +174,20 @@ def get_normalized_dataset(dataset, ps, seeds, d=0):
         s += seed + 1
         torch.manual_seed(s)
         x = torch.randn(p, d, dtype=torch.float64)
+
+        if stretching:
+            ds = d // 2 if params[0] is None else params[1]
+            x[:,ds:] *= stretching
```
Comment (mariogeiger) [inline dataset/__init__.py]: C'est normal (intuitif) que d-ds dimensions sont stretchée (au lieu de ds) ?

Comment (leonardopetrini) [reply dataset/__init__.py]: en fait c'est specifique au probleme du cylindre. Là on veut avoir ds dimension spherique et contrôler la variance des autres d-ds dimensions... On peut utilisé data_param2 pour specifier le dimensions à stretchée

Comment (leonardopetrini) [reply dataset/__init__.py]: ( il y a un typo, ...`else params[0]`, je corrige)

Comment (mariogeiger) [reply dataset/__init__.py]: là où tu as mis le code (pas dans le if de `cylinder`) ce n'est pas spécifique au cylindre

**Comment by leonardopetrini** [comment]
Oui, la raison pour laquelle on a d-ds dimensions stretchée au lieu de ds est spécifique au cylindre. Je l'avais mis dehors par ce que on pourrait l'utiliser pour les autres datasets aussi... Mais pour le moment je peut le mettre dans le if de cylinder pour créé pas de confusion. Si on fait comme ça je peux enlever aussi le paramètre stretching et mettre ça dans param[1] . Dis-moi si tu préfères comme ça !

Il 10 Mar 2020 11:01, Mario Geiger <notifications@github.com> ha scritto:

@mariogeiger commented on this pull request.

________________________________

In dataset/__init__.py<https://github.com/mariogeiger/feature_lazy/pull/8#discussion_r390204922>:

> @@ -168,11 +174,20 @@ def get_normalized_dataset(dataset, ps, seeds, d=0):

         s += seed + 1

         torch.manual_seed(s)

         x = torch.randn(p, d, dtype=torch.float64)

+

+        if stretching:

+            ds = d // 2 if params[0] is None else params[1]

+            x[:,ds:] *= stretching



là où tu as mis le code (pas dans le if de cylinder) ce n'est pas spécifique au cylindre

File: dataset/__init__.py (hunk 3 above)
Comment (leonardopetrini) [reply dataset/__init__.py]: Oui, la raison pour laquelle on a d-ds dimensions stretchée au lieu de ds est spécifique au cylindre. Je l'avais mis dehors par ce que on pourrait l'utiliser pour les autres datasets aussi... Mais pour le moment je peut le mettre dans le `if` de `cylinder` pour créé pas de confusion. Si on fait comme ça je peux enlever aussi le paramètre `stretching` et mettre ça dans` param[1] `. Dis-moi si tu préfères comme ça !

Comment (mariogeiger) [reply dataset/__init__.py]: J'aime bien l'idée de pouvoir stretcher tous les dataset mais dans ce cas il faut des paramètres intuitifs. Moi je mettrais des paramètres intuitifs et quand tu utilise le cylindre c'est à toi de penser à l'envers.

**Comment by mariogeiger** [comment]
Il y a des conflicts ! il faut que tu pull les derniers commit et que tu merge

**Comment by leonardopetrini** [comment]
> Il y a des conflicts ! il faut que tu pull les derniers commit et que tu merge

Resolved!

File: dataset/__init__.py [hunk 4]
```diff
@@ -173,6 +181,12 @@ def get_normalized_dataset(dataset, ps, seeds, d=0):
         if dataset == 'sphere':
             r = x.norm(dim=1)
             y = (r**2 > d - 2 / 3)
+        if dataset == 'cylinder':
+            dsph = d // 2 if params[0] is None else params[0]
+            stretching = 1 if params[1] is None else params[1]
```
Comment (mariogeiger) [inline dataset/__init__.py]: Je te conseil pas d'avoir des paramètres par défaut ici
```suggestion
            dsph = params[0]
            stretching = params[1]
            assert dsph is not None and stretching is not None
```

File: dataset/__init__.py [hunk 5]
```diff
@@ -182,6 +196,41 @@ def get_normalized_dataset(dataset, ps, seeds, d=0):
             y = (x[:, 0] > 0) * (x[:, 1] > 0)
         if dataset == 'andD':  # multi-dimensional AND logic gate (all d dimensions are relevant)
             y = (x > 0).all(1)
+        if dataset == 'sphere_grid':
+            assert d == 2, "Spherical grid is only implemented in 2D"
+            bins = 500 if params[0] is None else params[0]
+            theta_bins = 50 if params[1] is None else params[1]
```
Comment (mariogeiger) [inline dataset/__init__.py]: même chose ici
```suggestion
            bins = params[0]
            theta_bins = params[1]
            assert bins is not None and theta_bins is not None
```

File: dataset/__init__.py [hunk 6]
```diff
@@ -182,6 +196,41 @@ def get_normalized_dataset(dataset, ps, seeds, d=0):
...
+            for i in range(bins):
+                theta = (torch.rand(ppc) + (i % theta_bins)) / theta_bins * 2 * math.pi
+                r = torch.rand(ppc) * r_diff[i // theta_bins] + r_spacing[i // theta_bins]
+                x[i * ppc:(i + 1) * ppc, 0] = r.mul(theta.cos())
+                x[i * ppc:(i + 1) * ppc, 1] = r.mul(theta.sin())
+            r = x.norm(dim=1)
+            y = r > r_spacing[len(r_spacing) // 2]
+
+        if dataset == 'signal_1d':
+            n0 = 1 if params[0] is None else params[0]
```
Comment (mariogeiger) [inline dataset/__init__.py]: même chose ici

File: main.py [hunk 7]
```diff
@@ -465,6 +468,14 @@ def main():
     parser.add_argument("--pte", type=int)
     parser.add_argument("--d", type=int)
     parser.add_argument("--whitening", type=int, default=1)
+    parser.add_argument("--data_param1", type=int, default=None, help=
```
Comment (mariogeiger) [inline main.py]: C'est déjà `None` par default 
```suggestion
    parser.add_argument("--data_param1", type=int, help=
```

C'est toujours un int ? Peu être que mettrait un float, par exemple `stretching` ne peut pas être float ?
```suggestion
    parser.add_argument("--data_param1", type=float, default=None, help=
```
Et puis quand tu récupère tu fais `bins = int(param[0])`

**Review by mariogeiger — COMMENTED** [review]
Les arguments par défaut caché dans le code c'est pas bien pour se souvenir de quel argument a été utilisés quand on a lancé l'expériences. tous les arguments qui sont arbitraire ne devraient pas avoir de valeur par défaut.

File: main.py (hunk 7 above)
Comment (leonardopetrini) [reply main.py]: oui, bien sur, j'ai réalisé !

**Comment by mariogeiger** [comment]
Je merge ?

**Comment by leonardopetrini** [comment]
> Je merge ?

pour moi, oui !
